### PR TITLE
feat: add context menu to copy/paste using only the mouse

### LINKED ISSUE #211: Enable copy/paste right click on text
It'd be good to enable users to copy/paste by opening the traditional right click menu.

https://github.com/electron/electron/issues/4068#issuecomment-170911307
https://github.com/sindresorhus/electron-context-menu

### PR DESCRIPTION
**Pull request by tuliomir**
### Acceptance Criteria
- Shows a context menu on mouse right click on all text input fields allowing copy and paste
- Does not allow copy on password fields
- Resolves #211 

### Security Checklist
- [X] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

 ### UI/UX

https://user-images.githubusercontent.com/3298774/150235177-42f953f5-4b95-41d1-97fc-3f590820a36f.mov

### CODE REVIEW

File: public/contextMenu.js [hunk 1]
```diff
@@ -0,0 +1,409 @@
+/*
+MIT License
+
+Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (https://sindresorhus.com)
+Original from: https://github.com/sindresorhus/electron-context-menu
```
Comment (luislhl) [inline public/contextMenu.js:5]: Maybe we should make it explicit (here or in some other place) the version of the original on which ours is based.

This will make it easier if we need to update the code with newer versions

File: public/contextMenu.js [hunk 2]
```diff
@@ -0,0 +1,409 @@
+/*
```
Comment (luislhl) [inline public/contextMenu.js:1]: Did you make any changes from the original?

File: public/contextMenu.js [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+    Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+    WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+    OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+    OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+const electron = require('electron');
```
Comment (pedroferreira1) [inline public/contextMenu.js:20]: I think we should keep only the minimum required for our feature and not the whole file. Besides that, would be interesting to add support for `debugMode`. We have this in our `public/electron.js` and this could add the 'Inspect element' in the context menu.

**Review by pedroferreira1 — CHANGES_REQUESTED** [review]

File: public/contextMenu.js (hunk 2 above)
Comment (tuliomir) [reply public/contextMenu.js:1]: Yes, I removed the features that requires new dependencies to be installed.
Basically commented lines 21 to 23 and then commented every code instance that used them.

File: public/contextMenu.js (hunk 3 above)
Comment (tuliomir) [reply public/contextMenu.js:20]: The features we did not use were now removed from our code.

The overall library architecture was kept despite being a bit over-sized for our needs because it is already a stable code, handles edge cases and has good maintenance.

File: public/contextMenu.js (hunk 1 above)
Comment (tuliomir) [reply public/contextMenu.js]: Added a reference to the specific release we based our code on on the url itself.

File: public/electron.js [hunk 4]
```diff
@@ -41,6 +42,18 @@ const debugMode = (
   process.argv.indexOf('--debug') >= 0
 );
 
+// Creates the context menu on mouse right-click for allowing copy/paste and other configurable actions
+contextMenu({
+  menu: (actions, props, browserWindow, dictionarySuggestions) => {
+    return [
+        ...dictionarySuggestions,
```
[the remark this answers is not in the document]
Comment (tuliomir) [reply public/electron.js]: That was the recommended way to call this library, as per the author's `readme` file.

It was one of the features removed on the code cleanup, this function call is a lot more straightforward now.

File: public/contextMenu.js (hunk 2 above)
Comment (tuliomir) [reply public/contextMenu.js:1]: A new cleanup was made, removing many features we didn't need.
Our version is now significantly different from the original.

**Review by luislhl — APPROVED** [review]

File: public/contextMenu.js [hunk 5]
```diff
@@ -0,0 +1,258 @@
+/*
```
Comment (andreabadesso) [inline public/contextMenu.js:1]: **Comment unrelated to line number**

I know we don't have a linter on this project yet, but I think we should refactor this entire file to be more consistent with the code style from the other files by using 2 spaces indentation

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: public/contextMenu.js (hunk 5 above)
Comment (tuliomir) [reply public/contextMenu.js:1]: Good point! The file was refactored.

File: public/contextMenu.js [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    .filter((menuItem, index, array) => {
+      const toDelete = menuItem.type === 'separator' && (!notDeletedPreviousElement || index === array.length - 1 || array[index + 1].type === 'separator');
+      notDeletedPreviousElement = toDelete ? notDeletedPreviousElement : menuItem;
+      return !toDelete;
+    });
+};
+
+const create = (win, options) => {
+  const handleContextMenu = (event, props) => {
+    if (typeof options.shouldShowMenu === 'function' && options.shouldShowMenu(event, props) === false) {
```
Comment (pedroferreira1) [inline public/contextMenu.js:48]: What's this `options.shouldShowMenu`? I think we don't use it.

File: public/contextMenu.js [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+    const {editFlags} = props;
+    const hasText = props.selectionText.trim().length > 0;
+    const can = type => editFlags[`can${type}`] && hasText;
+
+    const defaultActions = {
+      separator: () => ({type: 'separator'}),
+      cut: decorateMenuItem({
+        id: 'cut',
+        label: 'Cu&t',
```
Comment (pedroferreira1) [inline public/contextMenu.js:60]: What's this &?

File: public/contextMenu.js [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        enabled: can('Cut'),
+        visible: props.isEditable,
+        click(menuItem) {
+          const target = webContents(win);
+
+          if (!menuItem.transform && target) {
```
Comment (pedroferreira1) [inline public/contextMenu.js:66]: What's this `if` doing?

File: public/contextMenu.js [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+    let dictionarySuggestions = [];
+
+    let menuTemplate = [
+      defaultActions.cut(),
+      defaultActions.copy(),
+      defaultActions.paste(),
+      defaultActions.separator(),
+      shouldShowInspectElement && defaultActions.inspect(),
+      defaultActions.separator()
```
Comment (pedroferreira1) [inline public/contextMenu.js:130]: If don't should the inspect element, then show two separators in a row?

File: public/contextMenu.js [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+    ];
+
+    if (options.menu) {
```
Comment (pedroferreira1) [inline public/contextMenu.js:133]: The only options parameter we are using is `showInspectElement`, right? Should we remove all those `if`'s?

File: public/contextMenu.js [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+    if (options.append) {
+      const result = options.append(defaultActions, props, win, event);
+
+      if (Array.isArray(result)) {
+        menuTemplate.push(...result);
+      }
+    }
+
+    menuTemplate = removeUnusedMenuItems(menuTemplate);
```
Comment (pedroferreira1) [inline public/contextMenu.js:153]: What's this doing?

File: public/contextMenu.js [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+    for (const menuItem of menuTemplate) {
```
Comment (pedroferreira1) [inline public/contextMenu.js:155]: Are we using this?

File: public/contextMenu.js (hunk 7 above)
Comment (tuliomir) [reply public/contextMenu.js:60]: & is a syntax for the context menu to know where to put the underscore on the word. And then, what "hotkey" will be assigned to that button.

In that case, a right click shows the context menu and if the user types `T` on the keyboard, the system runs the "Cut" operation.

File: public/contextMenu.js (hunk 6 above)
Comment (tuliomir) [reply public/contextMenu.js:48]: It is an optional function to allow the context menu to understand if, under runtime conditions, it should not be drawn.

We don't use it, I will remove it as well.

File: public/contextMenu.js (hunk 8 above)
Comment (tuliomir) [reply public/contextMenu.js:66]: Another generic structure that allows custom behavior when the operation is executed.
It is out of our scope, I'll remove it.

File: public/contextMenu.js (hunk 11 above)
Comment (tuliomir) [reply public/contextMenu.js:153]: The lib was structured in a way that it first includes all menuItems and then runs a `can` function to remove those that are not relevant for the runtime.

Since we won't handle a dynamic amount of menuItems we could do it the other way around, checking if the elements are needed before adding, but I don't think it will improve the code in a meaningful way.

I'll add some inline documentation to improve the code readability.

File: public/contextMenu.js (hunk 12 above)
Comment (tuliomir) [reply public/contextMenu.js:155]: No, removing this section as well.

File: public/contextMenu.js (hunk 9 above)
Comment (tuliomir) [reply public/contextMenu.js]: The "removeUsedMenuItems" had a logic to avoid two consecutive separators or a separator as a first/last element.
Since we also removed dynamic menuItems, this was further simplified.

**Comment by tuliomir** [comment]
@pedroferreira1 , a complete jsdoc was implemented on the contextMenu library and its scope was further restricted to our current use of Copy/Paste/Cut.
This should help with code clarity.

File: public/contextMenu.js (hunk 10 above)
Comment (tuliomir) [reply public/contextMenu.js]: Removed :+1:

File: public/contextMenu.js [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+
+const webContents = win => win.webContents || (win.id && win);
+
+/**
+ * Removes menuItems that are not being used
+ * @param menuTemplate
+ * @returns {*}
+ */
+const removeUnusedMenuItems = menuTemplate => {
+  let notDeletedPreviousElement;
```
Comment (pedroferreira1) [inline public/contextMenu.js:30]: This is not used, right? Should we delete it?

**Review by pedroferreira1 — APPROVED** [review]
